finality-aleph/src/sync/forest/mod.rs

```diff
@@ -289,52 +306,107 @@ impl<I: PeerId, J: Justification> Forest<I, J> {
+    pub fn try_finalize(&mut self, number: &u32) -> Option<J> {
+        // cached as ready
+        if let Some(id) = self.justified_blocks.get(number) {
+            if let Some(VertexWithChildren { vertex, children }) = self.vertices.remove(id) {
                 match vertex.ready() {
+                    // ready indeed
                     Ok(justification) => {
-                        self.root_id = child_id;
+                        self.root_id = id.clone();
                         self.root_children = children;
                         self.prune_level(self.root_id.number());
                         return Some(justification);
                     }
+                    // cache corrupted, we'll wait for it to fix itself
                     Err(vertex) => {
                         self.vertices
-                            .insert(child_id, VertexWithChildren { vertex, children });
+                            .insert(id.clone(), VertexWithChildren { vertex, children });
                     }
                 }
             }
         }
         None
     }
 
-    /// How much interest we have for the block.
-    pub fn state(&mut self, id: &BlockIdFor<J>) -> Interest<I> {
+    /// Returns the BranchKnowledge regarding the given block id,
+    /// or None if any errors are encountered.
+    fn branch_knowledge(&mut self, mut id: BlockIdFor<J>) -> Option<BranchKnowledge<J>> {
+        use VertexHandle::*;
+        let mut maybe_parent_id;
+        // traverse ancestors till we reach something imported or a parentless vertex
+        // avoid infinite loop by limiting the number of iterations to max possible value
+        for _ in 0..MAX_DEPTH {
+            // try get parent id
+            maybe_parent_id = match self.get_mut(&id) {
+                Candidate(entry) => {
+                    // first encounter of an imported ancestor
+                    if entry.get().vertex.imported() {
+                        return Some(BranchKnowledge::TopImported(id));
+                    }
+                    entry.get().vertex.parent().cloned()
+                }
+                HighestFinalized => {
+                    return Some(BranchKnowledge::TopImported(id));
+                }
+                // we shouldn't have reached any other type of vertex
+                _ => return None,
+            };
+            // try update current id to parent_id
+            match maybe_parent_id {
+                Some(parent_id) => id = parent_id,
+                // does not have parent, thus is not imported (as is not HighestFinalized)
+                None => return Some(BranchKnowledge::LowestId(id)),
+            };
+        }
+        None
+    }
+
+    /// Prepare additional info required to create a request for the block.
+    /// Returns `None` if the block is not required, its ID is not known, is a hopeless fork,
+    /// its number is out of allowed range, or if any other error was encountered.
```

Comment:
This really sounds as if it should return these errors in some way, rather than just describe them in the comment. Although changing the comment to be about returning `None` whenever we are not actually interested in the vertex, without mentioning the exact reasoning, might also be acceptable.

Reply:
changed the comments, not sure how the errors would help, as we would ignore them anyway in method `state`